pkg/gpu/nvidia/mig/mig.go

```diff
@@ -0,0 +1,357 @@
+	nvml.GPU_INSTANCE_PROFILE_2_SLICE: 3,
+	nvml.GPU_INSTANCE_PROFILE_3_SLICE: 2,
+	nvml.GPU_INSTANCE_PROFILE_4_SLICE: 1,
+	nvml.GPU_INSTANCE_PROFILE_7_SLICE: 1,
+}
+
+var partitionSizeToProfileMap = map[string]int{
+	"1g.5gb":  nvml.GPU_INSTANCE_PROFILE_1_SLICE,
+	"2g.10gb": nvml.GPU_INSTANCE_PROFILE_2_SLICE,
+	"3g.20gb": nvml.GPU_INSTANCE_PROFILE_3_SLICE,
+	"7g.40gb": nvml.GPU_INSTANCE_PROFILE_7_SLICE,
+}
+
+// DeviceManager performs various management operations on mig devices.
+type DeviceManager struct {
+	devDirectory  string
+	procDirectory string
+	devices       []*nvml.Device // Represents whole GPUs
+	gpuPartitions map[string][]pluginapi.DeviceSpec
+}
+
+// NewDeviceManager creates a new DeviceManager to handle MIG devices on the node.
+func NewDeviceManager(devDirectory, procDirectory string) DeviceManager {
+	return DeviceManager{
+		devDirectory:  devDirectory,
+		procDirectory: procDirectory,
+		devices:       make([]*nvml.Device, 0),
+		gpuPartitions: make(map[string][]pluginapi.DeviceSpec),
+	}
+}
+
+// ListGPUPartitionDevices lists all the GPU partitions as devices that can be advertised as
+// resources available on the node.
+func (d *DeviceManager) ListGPUPartitionDevices() map[string]pluginapi.Device {
+	devices := make(map[string]pluginapi.Device)
+
+	for id := range d.gpuPartitions {
+		devices[id] = pluginapi.Device{
+			ID:     id,
+			Health: pluginapi.Healthy,
+		}
+	}
+
+	return devices
+}
+
+// DeviceSpec returns the device spec that inclues list of devices to allocate for a deviceID.
+func (d *DeviceManager) DeviceSpec(deviceID string) ([]pluginapi.DeviceSpec, error) {
+	deviceSpecs, ok := d.gpuPartitions[deviceID]
+	if !ok {
+		return []pluginapi.DeviceSpec{}, fmt.Errorf("invalid allocation request with non-existing GPU partition: %s", deviceID)
+	}
+
+	return deviceSpecs, nil
+}
+
+// Start method performs the necessary initializations and starts the mig.DeviceManager.
+func (d *DeviceManager) Start() error {
+	if err := nvml.Init(); err != nil {
+		return fmt.Errorf("unable to initialize nvml: %v", err)
+	}
+
+	gpuCount, err := nvml.GetDeviceCount()
+	if err != nil {
+		return fmt.Errorf("failed to get device count: %v", err)
+	}
+
+	for i := uint(0); i < gpuCount; i++ {
+		device, err := nvml.NewDevice(i)
+		if err != nil {
+			return fmt.Errorf("failed to look up device index: %d, error: %v", i, err)
+		}
+
+		d.devices = append(d.devices, device)
+	}
+	return nil
+}
+
+// Stop method performs necessary cleanup operations to safely shut down mig.DeviceManager.
+func (d *DeviceManager) Stop() error {
+	return nvml.Shutdown()
+}
+
+// ConfigureGPUPartitions accepts a string describing the GPU partition type, and partitions all the GPUs on the node as specified.
+func (d *DeviceManager) ConfigureGPUPartitions(partition string) error {
+	if err := d.enableMIGMode(); err != nil {
+		return fmt.Errorf("failed to configure GPU partitions: %v", err)
+	}
+
+	if err := deleteAllGPUPartitions(); err != nil {
```

Comment:
is this needed for device plugin restarts, where we need to clean up the existing partitions first?

Reply:
Good point. Ideally if the GPU is already partitioned the way we expect, then we should just skip that step. I'll update this.